libraries/_config.rb

```diff
@@ -0,0 +1,218 @@
+#
+# Cookbook:: syslog_ng
+# Library:: common
+#
+# Copyright:: 2020, Ben Hughes <bmhughes@bmhughes.co.uk>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module SyslogNg
+  module Cookbook
+    SYSLOG_NG_PARAMETER_BUILD_TYPES ||= %i(source_dest rewrite).freeze
+    SYSLOG_NG_PARAMETERS_UNNAMED ||= { rewrite: %w(match replacement field tags additional_options) }.freeze
+    SYSLOG_NG_BOOLEAN_OPERATORS ||= %w(and or and_not or_not container).freeze # The __container__ 'operator' and __operator__ 'filter' are added here to allow nested boolean operation, syslog-ng doesn't know anything about it.
+    SYSLOG_NG_FILTER_FUNCTIONS ||= %w(facility filter host inlist level priority match message netmask netmask6 program source tags operator).freeze
+    SYSLOG_NG_REWRITE_OPERATORS ||= %w(subst set unset groupset groupunset set-tag clear-tag credit-card-mask credit-card-hash).freeze
+    SYSLOG_NG_FORMATTED_PARAMETER_REGEX ||= /(('|").*('|"))|(.+\(.+\))|(^(\d+ ?)+$)/.freeze
+
+    module ConfigHelpers
+      def format_parameter_value(value)
+        log_chef(:debug, "Receieved value to format: #{value} as #{value.class}.")
+        if value.is_a?(String)
+          if %w(yes YES no NO).include?(value) || value.match?(SYSLOG_NG_FORMATTED_PARAMETER_REGEX) || ip_address?(value)
+            log_chef(:debug, "Returning unformatted value string '#{value}'.")
+            return value
+          else
+            formatted = "\"#{value}\""
+            log_chef(:debug, "Formatted parameter value to: #{formatted}}.")
+            formatted
+          end
+        else
+          log_chef(:debug, "Returning unformatted parameter value '#{value}' as #{value.class}.")
+          value
+        end
+      end
+
+      def build_parameter_string(type, parameters)
```

Comment:
You, and the end user might be better off chose either a hash, Array or String, rather than all three. This will significantly reduce the complexity of using and maintaining this.

Reply:
This ties in to the usage in `log` below and also to allow it to deal with nested hashes and arrays. I'll agree it's not something you can maintain without getting into it but we'd have to remove a lot of functionality otherwise 😕